source-academy/js-slang
# Add Source 1 Typed variant

Changes:
-
- Added ability to parse TS code to ESTree AST format, using a combination of [Babel Parser](https://babeljs.io/docs/en/babel-parser) and [acorn Typescript plugin](https://github.com/patternfly/patternfly-org/blob/main/packages/ast-helpers/acorn-typescript.js).
- Added a type error checker, which recurses through the parsed AST and adds any detected type errors to the context. The type error checker is designed such that all types are assumed to be "any" unless specifically declared (e.g. declared by the user using type annotations, pre-declared Source 1 prelude functions, binary operations such as "+" being restricted to only strings and numbers in Source 1, etc). Note that this type error checker is separate from the existing type checker used in type inference, however functions shared between the two checkers have been abstracted out into a separate utils file.
- Added types used in the type error checker. The types for the types are mostly the same as those used in the type inference checker, but additional ones have been added as per necessary and certain existing types were renamed for clarity.
- Modified the Source mode for the frontend Ace Editor to support syntax highlighting differences in the Source Typed variants.
- Added unit tests to cover most of Source 1 Typed functionality.

How to Test this PR:
-
`src/parser/typeParser.ts` can be ignored as it is the code for the [acorn Typescript plugin](https://github.com/patternfly/patternfly-org/blob/main/packages/ast-helpers/acorn-typescript.js). This reduces about 1k LoC of code to review, but that still makes this PR about 3k LoC large, apologies in advance.

Given the size of the PR, the main focus would therefore be to test the functionality of Source Typed to see if there are any bugs/inconsistencies. This can be done in the following ways:
1. Without frontend: Run Source 1 Typed in the js-slang REPL.
```
$ js-slang -c 1 -v typed # if js-slang is added to PATH
$ node dist/repl/repl.js -c 1 -v typed # if js-slang is not added to PATH
```
2. With frontend: [Link local js-slang to local frontend](https://github.com/source-academy/js-slang#using-your-js-slang-in-your-local-source-academy) and checkout local frontend to source-academy/frontend#2264.
3. Check the existing unit tests in `src/typeChecker/__tests__/typeErrorChecker.test.ts`.
```
$ yarn test src/typeChecker/__tests__/typeErrorChecker.test.ts
```
